src/constraint_system/boolean.rs

```diff
@@ -57,27 +57,32 @@ where
     pub fn is_zero_gate(&mut self, a: Variable) -> Variable {
         //Get relevant field values
         let a_value = self.variables.get(&a).unwrap();
-        let a_inv_value = a_value.inverse()
-            .unwrap_or(E::Fr::one());
+        let a_inv_value = a_value.inverse().unwrap_or(E::Fr::one());
         //This has value 1 if input value is zero, value 0 otherwise
         let result_value = E::Fr::one() - *a_value * a_inv_value;
 
         let a_inv = self.add_input(a_inv_value);
         let result = self.add_input(result_value);
 
         // Enforce constraints
-        let a_times_result = self.arithmetic_gate(|gate| gate.witness(a, result, None).mul(E::Fr::one()));
+        let a_times_result = self.arithmetic_gate(|gate| {
+            gate.witness(a, result, None).mul(E::Fr::one())
+        });
         self.assert_equal(a_times_result, self.zero_var());
 
-        let a_times_ainv = self.arithmetic_gate(|gate| gate.witness(a, a_inv, None).mul(E::Fr::one()));
-        let sum = self.arithmetic_gate(|gate| gate.witness(a_times_ainv, result, None).add(E::Fr::one(), E::Fr::one()));
+        let a_times_ainv = self.arithmetic_gate(|gate| {
```

Comment:
The gates beginning here at line 73 can all be squashed into a single gate if you enable fan_in_3. 

`self.arithmetic_gate(|gate| {
    gate.witness(a, a_inv, self.zero_var())
        .mul(E::Fr::one())
        .fan_in_3(E::Fr::one(), result)
        .constant(-E::Fr:one())
`
This single gate constrains `1 * a * a_inv + 1 * result  - 1 = 0` and also avoids using another unnecessary `assert_equal` gate in line 81.

If you don't want to use `fan_in_3` you can do this in two gates by keeping the `mul` gate as you had it, and putting `self.zero_var()` as the output variable in the following `add` gate and still avoid the unnecessary `assert_equal` gate.